### ISSUE TITLE
[BUG] longhorn-manager process is consuming high memory on nodes

### ISSUE THREAD

**Issue by plusiv** [longhorn/longhorn#6645]
## Describe the bug (🐛 if you encounter this issue)

The longhorn-manager process is consuming an excessive amount of RAM in multiple nodes.

<img width="1026" alt="image" src="https://github.com/longhorn/longhorn/assets/37001774/f20d25b3-1138-4200-9630-86ace7eeab2e">

## To Reproduce

I have 13 nodes but just 7 of them are schedulable for storage. All schedulable nodes have 500 GiB of SSD dedicated disk, 8 cores and 24 GiB of Memory.

## Expected behavior

Memory consumption being relatively optimal.

## Support bundle for troubleshooting

[Download from drive](https://drive.google.com/file/d/15pgjuuIpk6HEM5-HRMvI0KNfVdBKAzkV/view?usp=sharing)

## Environment

 - Longhorn version: v1.5.1
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Rancher Catalog
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: RKE2
   - Number of management node in the cluster: 4
   - Number of worker node in the cluster: 13
 - Node config
   - OS type and version: Ubuntu 22.04.3 LTS
   - Kernel version: 5.15.0-79-generic
   - CPU per node: Min. 8 cores
   - Memory per node: Min 16 GiB
   - Disk type(e.g. SSD/NVMe/HDD): SSD
   - Network bandwidth between the nodes: 1 Gibps
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): VMWare
 - Number of Longhorn volumes in the cluster: 16
 - Impacted Longhorn resources: N/A
   - Volume names:

## Additional context
The used tool for diagnose was [htop](https://htop.dev/).

**Comment by ChanYiLin** [comment]
I saw there was a replica of a volume `pvc-ffac8692-1f38-40f2-a8a9-4479c0f8a9d6` kept failing and restarting
Did you find any Volume having abnormal or issues?

**Comment by plusiv** [comment]
Hi @ChanYiLin , thank you for your response. 

I faced similar issues of volumes having abnormals behaviors (replicas falling and restarting), however it was caused by a misconfiguration error about `Storage Over Provisioning Percentage` in `200` and `longhorn-manager` pods with poor resources assigned in their deployments. After that recurrent falling and restarting behavior was solved, but high memory usage still persisting.

I have a few months trying to troubleshoot that issue from longhorn.

**Comment by shuo-wu** [comment]
We are sorry for not getting back to you in time. Here are some questions & findings:
1. Would you provide the longhorn components resource usage info? `kubectl -n longhorn-system top pod`
2. Which components is the high memory usage caused by? longhorn-manager pods, or instance-manager pods?
3. Do the problematic longhorn components **keep using lots of memory**? Is there a metric that presents the resource usage change?
4. In the instance manager `instance-manager-c86e5a7c0c3a673b2dbfea5d9fe4dc35` I found tons of error log `keepalive ping failed to receive ACK within timeout`. I guess this is one cause. And this is already discussed in https://github.com/longhorn/longhorn/issues/6578#issuecomment-1718671394
```
2023-09-06T19:33:25.010330738Z [longhorn-instance-manager] time="2023-09-06T19:33:25Z" level=error msg="Failed to receive next item in process watch" error="rpc error: code = Unavailable desc = keepalive ping failed to receive ACK within timeout"
```
5. There are some rebuildings that could contribute **temporary**/**long-term** CPU & memory usage surge.
    1. In some instance manager pods, I found that sometimes there are some timeouts between engines and their replicas, which then triggers replica rebuilding.
    2. In some longhorn-manager pods, there are some **nodes becoming NotReady** cases, which may indicate the node down then lead to the instance manager pods crashing. An instance manager pod crashing means each replica on this node will experience a rebuilding later. Just in case, I think you can check the cluster stability first.
    ```
    2023-08-17T15:03:31.678571360Z time="2023-08-17T15:03:31Z" level=info msg="Instance manager state is updated from unknown to error after syncing with the pod" controller=longhorn-instance-manager instanceManager=instance-manager-e7e7f976ddb1c99865c1d56f86dc8025 node=worker04 nodeID=worker04

    2023-08-18T03:00:29.735771917Z time="2023-08-18T03:00:29Z" level=info msg="Event(v1.ObjectReference{Kind:\"Node\", Namespace:\"longhorn-system\", Name:\"worker12\", UID:\"30fc0550-343b-431d-b4c9-85b931ddeb6a\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"229798865\", FieldPath:\"\"}): type: 'Warning' reason: 'Ready' Kubernetes node worker12 not ready: KubeletNotReady"
    2023-08-18T03:00:29.747983683Z time="2023-08-18T03:00:29Z" level=info msg="Event(v1.ObjectReference{Kind:\"Node\", Namespace:\"longhorn-system\", Name:\"worker12\", UID:\"30fc0550-343b-431d-b4c9-85b931ddeb6a\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"230148352\", FieldPath:\"\"}): type: 'Normal' reason: 'Ready' Node worker12 is ready"
    2023-08-18T03:00:29.747996655Z time="2023-08-18T03:00:29Z" level=info msg="Event(v1.ObjectReference{Kind:\"Node\", Namespace:\"longhorn-system\", Name:\"worker12\", UID:\"30fc0550-343b-431d-b4c9-85b931ddeb6a\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"230148352\", FieldPath:\"\"}): type: 'Warning' reason: 'Ready' Kubernetes node worker12 not ready: KubeletNotReady"
    2023-08-18T03:00:29.754621216Z time="2023-08-18T03:00:29Z" level=info msg="Event(v1.ObjectReference{Kind:\"Node\", Namespace:\"longhorn-system\", Name:\"worker12\", UID:\"30fc0550-343b-431d-b4c9-85b931ddeb6a\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"230148352\", FieldPath:\"\"}): type: 'Normal' reason: 'Ready' Node worker12 is ready"
    ```
    3. Especially in instance manager pod `instance-manager-e7e7f976ddb1c99865c1d56f86dc8025`, there are lots of replica rebuilding log. You can check if this instance manager pod has eaten lots of resources in the past few days.

**Comment by derekbit** [comment]
@plusiv 
After discussing with @shuo-wu, we've fixed the instance-manager pod with high cpu usage. If you're available, we can provide the custom instance-manager image. We can see if there is any potential issue after upgrade.

**Comment by plusiv** [comment]
Hi everyone, happy to read feedback!

@shuo-wu  Regarding to
> 1. Would you provide the longhorn components resource usage info? kubectl -n longhorn-system top pod
> 2. Which components is the high memory usage caused by? longhorn-manager pods, or instance-manager pods?
> 3. Do the problematic longhorn components keep using lots of memory? Is there a metric that presents the resource usage change?
> 4. In the instance manager instance-manager-c86e5a7c0c3a673b2dbfea5d9fe4dc35 I found tons of error log keepalive ping failed to receive ACK within timeout. I guess this is one cause. And this is already discussed in https://github.com/longhorn/longhorn/issues/6578#issuecomment-1718671394

1. Got the following output
```shell
> kubectl -n longhorn-system top pod
NAME                                                     CPU(cores)   MEMORY(bytes)   
csi-attacher-84575b8699-gzc7r                            1m           17Mi            
csi-attacher-84575b8699-mj82h                            1m           13Mi            
csi-attacher-84575b8699-n8k7b                            1m           12Mi            
csi-provisioner-6cdc4c5f4-5nz2v                          1m           22Mi            
csi-provisioner-6cdc4c5f4-kfpmx                          1m           12Mi            
csi-provisioner-6cdc4c5f4-l5k7p                          1m           9Mi             
csi-resizer-7468668746-b9gsb                             1m           10Mi            
csi-resizer-7468668746-wvjct                             1m           9Mi             
csi-resizer-7468668746-xzrd5                             1m           20Mi            
csi-snapshotter-5b9ffcd798-8ht8c                         1m           10Mi            
csi-snapshotter-5b9ffcd798-8xmvv                         1m           14Mi            
csi-snapshotter-5b9ffcd798-tsrm4                         1m           22Mi            
engine-image-ei-65af8925-584zd                           12m          22Mi            
engine-image-ei-65af8925-6d46p                           11m          20Mi            
engine-image-ei-65af8925-9722w                           12m          22Mi            
engine-image-ei-65af8925-b7rsp                           11m          21Mi            
engine-image-ei-65af8925-cdqwg                           10m          21Mi            
engine-image-ei-65af8925-dbxjn                           11m          21Mi            
engine-image-ei-65af8925-fgfwq                           10m          16Mi            
engine-image-ei-65af8925-pkdkc                           11m          17Mi            
engine-image-ei-65af8925-pzx7p                           10m          20Mi            
engine-image-ei-65af8925-qh7v7                           11m          21Mi            
engine-image-ei-65af8925-rm2dc                           10m          19Mi            
engine-image-ei-65af8925-t97d5                           13m          23Mi            
engine-image-ei-65af8925-x8272                           11m          19Mi            
instance-manager-33dc7559de87682b9cff4cde1543a939        11m          41Mi            
instance-manager-5179180ce4ca577e8d5d74eb852a3863        5m           127Mi           
instance-manager-59e5d08c3eadc373dc4c5b311b641581        29m          84Mi            
instance-manager-661500d93241727bd17f5716ff304fa5        17m          63Mi            
instance-manager-6adb6f24812195f41e2fedd07094d6e2        1m           21Mi            
instance-manager-89299cadbbef0b40070855c22f52a939        53m          431Mi           
instance-manager-8e20e40b4c35991ec895385f87551cec        1245m        122Mi           
instance-manager-95bb41f4cb4acd2d6bf39d2004fcc9cf        1m           12Mi            
instance-manager-a2ce18f0fc3241304eb1754e993eba1a        26m          368Mi           
instance-manager-ae102c46e2b11e1088aad57c04cb29d3        26m          334Mi           
instance-manager-bf6f4da1a8f541496fbfe803934d6874        7m           46Mi            
instance-manager-c86e5a7c0c3a673b2dbfea5d9fe4dc35        1260m        136Mi           
instance-manager-e7e7f976ddb1c99865c1d56f86dc8025        13m          310Mi           
longhorn-csi-plugin-58jzn                                1m           29Mi            
longhorn-csi-plugin-74xmg                                1m           36Mi            
longhorn-csi-plugin-7hfhr                                1m           36Mi            
longhorn-csi-plugin-9jx87                                1m           41Mi            
longhorn-csi-plugin-d5sr8                                1m           44Mi            
longhorn-csi-plugin-f2ssr                                1m           38Mi            
longhorn-csi-plugin-grlng                                1m           47Mi            
longhorn-csi-plugin-h5stw                                1m           50Mi            
longhorn-csi-plugin-j54l7                                1m           40Mi            
longhorn-csi-plugin-kxnbd                                1m           51Mi            
longhorn-csi-plugin-p4msh                                1m           39Mi            
longhorn-csi-plugin-spw9q                                1m           39Mi            
longhorn-csi-plugin-tc5bd                                1m           32Mi            
longhorn-driver-deployer-7d8f9fcdf9-pqqsv                1m           12Mi            
longhorn-manager-2dl8s                                   9m           747Mi           
longhorn-manager-2wtff                                   9m           581Mi           
longhorn-manager-44pm4                                   11m          690Mi           
longhorn-manager-4dhct                                   13m          876Mi           
longhorn-manager-6t95k                                   9m           732Mi           
longhorn-manager-7n6pm                                   12m          780Mi           
longhorn-manager-99t5c                                   10m          638Mi           
longhorn-manager-c44jg                                   9m           744Mi           
longhorn-manager-cvg7x                                   12m          735Mi           
longhorn-manager-h4ldz                                   14m          652Mi           
longhorn-manager-k6fpb                                   6m           718Mi           
longhorn-manager-rttdw                                   8m           698Mi           
longhorn-manager-whtvf                                   11m          603Mi           
longhorn-ui-79fb87588b-sg9k6                             0m           11Mi            
share-manager-pvc-ccca50ff-7d8a-4685-993f-20d4f631f8a4   3m           39Mi
```
2. I've observed the behavior in the __longhorn-manager__ as described in the image above.
3. No. We're are facing several problems with metrics storage due this problem.

@derekbit regarding to this
> @plusiv
> After discussing with @shuo-wu, we've fixed the instance-manager pod with high cpu usage. If you're available, we can provide the custom instance-manager image. We can see if there is any potential issue after upgrade.

Sure. How can we test that?

**Comment by derekbit** [comment]
@plusiv 

Thank you. Let's update and monitor the pods' status.

- Installation or upgrade by kubectl
  https://github.com/longhorn/longhorn/blob/v1.5.1/deploy/longhorn.yaml#L4101
  Replace `https://github.com/longhorn/longhorn/blob/v1.5.1/deploy/longhorn.yaml#L4101` with `derekbit/longhorn-instance-manager:v1.5.1-keepalive`


- Installation or upgrade by helm
  https://github.com/longhorn/longhorn/blob/v1.5.1/chart/values.yaml#L41-L42)
  Replace 
  ```
    instanceManager:
      repository: longhornio/longhorn-instance-manager
      tag: v1.5.1
  ```
  with
  ```
    instanceManager:
      repository: derekbit/longhorn-instance-manager
      tag: v1.5.1-keepalive
  ```

**Comment by plusiv** [comment]
@derekbit done.

**Comment by derekbit** [comment]
cc @shuo-wu

**Comment by shuo-wu** [comment]
After the upgrade, if you still encounter the high-memory longhorn manager pods, would you pick up the longhorn-manager and export its pprof heap result?

1. Forward the pprof port for the longhorn manager:
```
kubectl -n longhorn-system port-forward <longhorn manager pod name> 6060:6060
```
2. Get the heap profiling file:
```
wget -O heap.out "http://localhost:6060/debug/pprof/heap"
```
3. Upload the profiling result to this ticket. Of course, you can directly check it as well
```
go tool pprof heap.out
(pprof) web
```

**Comment by derekbit** [comment]
Hello @plusiv
Do the longhorn-manager pods consume lower memories currently?

**Comment by plusiv** [comment]
Hi everyone, sorry for being late to respond. I was reviewing my nodes consumption, and I have no found any increase of memory by the `longhorn-manager` process, at the moment everything looks fine.

**Comment by lukaszraczylo** [comment]
@derekbit hit the same issue. Would you kindly rebuild the image with both amd64 & arm64 support please? :)

**Comment by derekbit** [comment]
> @derekbit hit the same issue. Would you kindly rebuild the image with both amd64 & arm64 support please? :)

@lukaszraczylo Sure. I've added both amd64 and arm64. Please check OS/ARCH in the following link.
https://hub.docker.com/layers/derekbit/longhorn-instance-manager/v1.5.1-keepalive/images/sha256-c1d55daf5527f48dbb8e152ff7e0a97bf062bc5b8c02b3fe82bd127a3360ebda?context=explore

The fix is already included in the upcoming v1.5.2. Please let me know if run into any issue. Thank you.

**Comment by derekbit** [comment]
> @derekbit hit the same issue. Would you kindly rebuild the image with both amd64 & arm64 support please? :)

Hello @lukaszraczylo 
Before applying the workaround, can you provide us a support bundle and the[ pprof profiling file](https://github.com/longhorn/longhorn/issues/6645#issuecomment-1759247638)? Thank you.

**Comment by derekbit** [comment]
@lukaszraczylo 
Can you provide more information of your cluster? Thank you.
- Number of nodes
- Kubernetes version
- Longhorn version
- OS distro
- Number of RWO and RWO volumes
- Number of replicas

**Comment by lukaszraczylo** [comment]
@derekbit
* 20 raspberry pi ( mix of 4gb & 8gb ), out of which 3 nodes with nvme disks attached for longhorn storage )
* kubernetes version: v1.27.6+k3s1
* longhorn version: 1.5.1
* OS: raspbian
* 38 volumes total, 20 RWO, 18 RWM
* 2 replicas per volume

**Comment by lukaszraczylo** [comment]
@derekbit update - I flipped the image to the keepalive version, and memory usage in the cluster dropped from 70% to 63% for the past 24 hours, so I'd call it a win. Longhorn manager processes still use 330M of RAM, but I'm pretty sure you'll come up with some extra fix :)

![Screenshot 2023-10-24 at 08 25 39](https://github.com/longhorn/longhorn/assets/2182556/691cc961-bea1-457f-bdf7-676310f51815)

**Comment by derekbit** [comment]
@lukaszraczylo 
Let's close the ticket.

We've filed PRs for the improvement of longhorn-manager memory usage. More information is in https://github.com/longhorn/longhorn/issues/6866#issuecomment-1777653170. All the improvements are included in the upcoming v1.5.2.

Feel free to reopen it if there is any question or feedback. Thank you.

**Comment by derekbit** [comment]
duplicate https://github.com/longhorn/longhorn/issues/6866
